lucene/analysis/common/src/java/org/apache/lucene/analysis/synonym/Word2VecSynonymFilter.java

```diff
@@ -0,0 +1,109 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.analysis.synonym;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import org.apache.lucene.analysis.TokenFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.synonym.SynonymProvider.WeightedSynonym;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionLengthAttribute;
+import org.apache.lucene.analysis.tokenattributes.TypeAttribute;
+import org.apache.lucene.search.BoostAttribute;
+
+/**
+ * Applies single-token synonyms from a Word2Vec trained network to an incoming {@link TokenStream}.
+ *
+ * @lucene.experimental
+ */
+public final class Word2VecSynonymFilter extends TokenFilter {
+
+  private final CharTermAttribute termAtt = addAttribute(CharTermAttribute.class);
+  private final PositionIncrementAttribute posIncrAtt =
+      addAttribute(PositionIncrementAttribute.class);
+  private final PositionLengthAttribute posLenAtt = addAttribute(PositionLengthAttribute.class);
+  private final BoostAttribute boostAtt = addAttribute(BoostAttribute.class);
+  private final TypeAttribute typeAtt = addAttribute(TypeAttribute.class);
+
+  private final SynonymProvider synonymProvider;
+  private final int maxResult;
+  private final float accuracy;
+  private final LinkedList<WeightedSynonym> outputBuffer = new LinkedList<>();
+  private State lastState = null;
+
+  /**
+   * Apply previously built synonymProvider to incoming tokens.
+   *
+   * @param input input tokenstream
+   * @param synonymProvider synonym provider
+   * @param maxResult maximum number of result returned by the synonym search
+   * @param accuracy minimal value of cosign similarity between the searched vector and the
+   *     retrieved ones
+   */
+  public Word2VecSynonymFilter(
+      TokenStream input, SynonymProvider synonymProvider, int maxResult, float accuracy) {
+    super(input);
+    this.synonymProvider = synonymProvider;
+    this.maxResult = maxResult;
+    this.accuracy = accuracy;
+  }
+
+  @Override
+  public boolean incrementToken() throws IOException {
+
+    if (!outputBuffer.isEmpty()) {
+      // We still have pending outputs from a prior synonym match:
+      releaseBufferedToken();
+      return true;
+    }
+
+    if (input.incrementToken()) {
+      String term = new String(termAtt.buffer(), 0, termAtt.length());
```

Comment:
not sure 'String' term is acceptable, we may need to use more low level implementation for terms